README
$KAME: README,v 1.3 2005/01/12 06:06:10 suz Exp $

kame-dhcp6

	kame-dhcp6 is an open-source implementation of Dynamic Host
	Configuration Protocol for IPv6 (DHCPv6) developed by the KAME
	project.

	The implementation mainly conforms to the following standards:

	- RFC3315: Dynamic Host Configuration Protocol for IPv6 (DHCPv6)
	- RFC3319: Dynamic Host Configuration Protocol (DHCPv6) Options 
         	   for Session Initiation Protocol (SIP) Servers
	- RFC3633: IPv6 Prefix Options for Dynamic Host Configuration
	           Protocol (DHCP)
	- RFC3646: DNS Configuration options for Dynamic Host
	           Configuration Protocol for IPv6 (DHCPv6)

	All the components of the main protocol are provided, i.e.,
	DHCPv6 clients, servers, and relay agents.

	Supported message types are as follows:
	  Solicit, Advertise, Request, Renew, Rebind, Release, Reply,
	  Information-request, Relay-forward and Relay-reply.

	Note that the current implementation does not support IPv6
	temporary address allocation by DHCPv6, and there is no plan to
	implement that feature at the moment.

Bug Reports and Mailing Lists

	Bugs reports should be put to the Bug Tracking System at 
	sourceforge.net
		http://sourceforge.net/projects/wide-dhcpv6/

	Information on the mailing list, including how to subscribe
	and browse archives, is available at the following URL:

		http://sourceforge.net/mail/?group_id=153668


CHANGES
2008-07-30  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6[cs].conf.sample: added several comments/examples.
	  (contributed by Stefan Sperling)

	--- 20080615 tar ball released ---
2008-05-01  SUZUKI, Shinsuke <suz@kame.net>
	* common.c: fixed a memory leak (Bug-ID: 1847587)
	* dhcp6s.c: fixed a potential NULL pointer access (Bug-ID: 1848304).
		    used a val_statefuladdr, instead of val_prefix, to access 
		    an address in IA in DHCP DECLINE handling.
	* config.c: fixed a possible memory leak. (Bug-ID 1847588)
	* dhcp6s.c: fixed a lifetime calculation failure in RENEW/REBIND 
		    process for stateful-address. (Bug-ID: 1850457)
	* common.c: fixed a improper handling of a domain-name ending with '.'.
		    (Bug-ID: 1852869)

2007-12-06  SUZUKI, Shinsuke <suz@kame.net>
	* common.c: fixed a possible memory leak. (Bug-ID 1844676)
	* common.c: fixed a possible double free. (Bug-ID 1844683)
	* dhcp6s.c: fixed a bugthat dhcp6s cannot accept a relayed
	  request message with authentication option. (Bug-ID 1814287)

2007-07-25  SUZUKI, Shinsuke <suz@kame.net>
	* config.c, lease.h, common.c: removed compilation warnings

2007-07-24  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6s.c, dhcp6c.c, dhcp6relay.c, configure*: use sig_atomic_t for
	  a signal-handler variable (Bug-ID 1678874)
	* common.c: fixed a buffer overrun in sprint_uint64()
	  (Bug-ID 1714648)

2007-07-10  SUZUKI, Shinsuke <suz@kame.net>
	* common.c: fixed a singular-point detection failure in replay check.
	  (Bug-ID 1714644)

2007-06-28  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6s.8: fixed a typo in a manual (program name "dhcp6sctl" should
	  be "dhcp6ctl")

	--- 20070507 tar ball released ---
2007-05-03  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6relay.c: changes the destination port of a relayed DHCP packet
	  from 546 to 547, when it is supposed to be bound for a relay,
	  instead of a client.  (the situation occurs when DHCPv6-relays are
	  cascased in a row)
	* dhcp6s.c: fixed a Solicitation message validation error (introduced 
	  on Mar 21 2007)

2007-03-23  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6relay.c,dhcp6.c: in signal-receiving procedure, not uses 
	a function that is not always designed to be signal-safe 
	(Bug-ID 1678874)

2007-03-21  SUZUKI, Shinsuke <suz@kame.net>
	fixed several bugs found in IPv6-Ready-Logo DHCPv6 Self-Test script.
	* dhcp6s.c, config.h, if.c: implemented DHCPv6 Confirm message processing
	on DHCPv6 server.
	* dhcp6s.c, lease.c: implemented DHCPv6 Decline message processing on DHCPv6 server.
	* dhcp6s.c: fixed a Solicitation message validation as stated in RFC3315 15.2.
	* dhcp6s.c: fixed a bug that Rebind does not work for IA-NA.
	* config.c: supported an address pool that has only one entry.

2007-03-20  SUZUKI, Shinsuke <suz@kame.net>
	* common.c: fixed a bug that IA-PD/IA-NA cannot coexist in one DHCP message
	  when they have the same IA-ID.

2007-02-28  SUZUKI, Shinsuke <suz@kame.net>
	* config.c: fixed a typo.

2007-02-27  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6relay.c, dhcp6relay_script.c, common.[ch], dhcp6c_script.c,
	  Makefile.in, dhcp6relay.8:
	  supported script execution for dhcp6relay (contributed by Bruno STEVANT)

	* Makefile.in, configure.in, configure, auth.h, common.c, control.h, dhcp6.h,
	  dhcp6s.c, missing/{arc4random.h, daemon.c, err.h, getifaddrs.c, ifaddrs.h,
	  sys/queue.h, warnx.c}
	  supported compilation on Solaris (contributed by James Carlson)

2007-02-12  SUZUKI, Shinsuke <suz@kame.net>
	* cftoken.l: accepts interface names, such as eth0.100(Linux VLAN interface)
	  or foobar (named by "ifconfig name" command).  (Bug-ID 1644637)
	* configure.in, Makefile.in: allows a user to specify the owner/group for the
	  installed files.  (suggested by Tsuyoshi MOMOSE and Tomofumi Hayashi)

2007-01-11  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6s.conf.5: fixed an incorrect descrption regarding NTP option.
2006-12-04  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6c.8: clarified the behavior of the Info-req mode.
	* addrconf.c, common.[ch], configure, configure.in, dhcp6relay.c,
	prefixconf.c: introduced a HAVE_SA_LEN macro to support OSes 
	(other than linux) without length field in sockaddr_*

2006-11-15  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6s.c: fixed an IA allocation failure from pool when it doubly
	requested by the same client.  (patch submitted by Hajimu UMEMOTO)
	* Makefile.in: creates man/man[58] in "make install" 
	(suggested by Masahide NAKAMURA)

2006-11-05  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6s.c: dhcp6s recognizes "-P" option.
	(patch submitted by Masahide NAKAMURA)

	* dhcp6relay.c: accepts ULA (Bug-ID 1584704)

	* Makefile.in: fixed a compilation error on SuSE and Debian Linux.
	(patch submitted by Masahide NAKAMURA)

	--- 20061016 tar ball released ---
2006-10-13  SUZUKI, Shinsuke <suz@kame.net>
	* Makefile.in: creates the target directory in "make install" 
	(suggested by Masahide NAKAMURA)

2006-10-08  SUZUKI, Shinsuke <suz@kame.net>
	* dhcp6s.conf.sample: corrected a misleading comment (pointed out by
	Hajimu UMEMOTO)

2006-10-04  SUZUKI, Shinsuke <suz@kame.net>
	* config.c: fixed a improper memory access when there is no free
	  address in an address pool.  (Bug-ID 1563232)

	  fixed a bug that a pooled address will never be reused even when
	  it is released.

2006-10-03  SUZUKI, Shinsuke <suz@kame.net>
	* config.c: dhcp6s does not include a 0 refresh-time option in 
	ADVERTISE/REPLY message, when "option refreshtime" configuration does 
	not exist in its configuration.  (patch submitted by Jun-ya Kato)

	* cftoken.l, cfparse.y: fixed a compilation failure by flex 2.5.4
	 (patch submitted by Jun-ya Kato)

2006-10-01  SUZUKI, Shinsuke <suz@kame.net>
	* cftoken.l, dhcp6[cs].conf.5: accepted "nis-domain-name" and "nisp-
	  domain-name" as a configuration keyword for NIS/NIS+ domain name,
	  as well as nis/nisp-server-domain-name. (Bug-ID 1566494)

	* cftoken.l: supported a declaration name (hostname, poolname,
	  keyname, authname) including "-".  (Bug-ID 1566476)


dhcp6.h
/*	$KAME: dhcp6.h,v 1.56 2005/03/20 06:46:09 jinmei Exp $	*/
/*
 * Copyright (C) 1998 and 1999 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef __DHCP6_H_DEFINED
#define __DHCP6_H_DEFINED

/* Error Values */
#define DH6ERR_FAILURE		16
#define DH6ERR_AUTHFAIL		17
#define DH6ERR_POORLYFORMED	18
#define DH6ERR_UNAVAIL		19
#define DH6ERR_OPTUNAVAIL	20

/* Message type */
#define DH6_SOLICIT	1
#define DH6_ADVERTISE	2
#define DH6_REQUEST	3
#define DH6_CONFIRM	4
#define DH6_RENEW	5
#define DH6_REBIND	6
#define DH6_REPLY	7
#define DH6_RELEASE	8
#define DH6_DECLINE	9
#define DH6_RECONFIGURE	10
#define DH6_INFORM_REQ	11
#define DH6_RELAY_FORW	12
#define DH6_RELAY_REPLY	13

/* Predefined addresses */
#define DH6ADDR_ALLAGENT	"ff02::1:2"
#define DH6ADDR_ALLSERVER	"ff05::1:3"
#define DH6PORT_DOWNSTREAM	"546"
#define DH6PORT_UPSTREAM	"547"

/* Protocol constants */

/* timer parameters (msec, unless explicitly commented) */
#define SOL_MAX_DELAY	1000
#define SOL_TIMEOUT	1000
#define SOL_MAX_RT	120000
#define INF_TIMEOUT	1000
#define INF_MAX_RT	120000
#define REQ_TIMEOUT	1000
#define REQ_MAX_RT	30000
#define REQ_MAX_RC	10	/* Max Request retry attempts */
#define REN_TIMEOUT	10000	/* 10secs */
#define REN_MAX_RT	600000	/* 600secs */
#define REB_TIMEOUT	10000	/* 10secs */
#define REB_MAX_RT	600000	/* 600secs */
#define REL_TIMEOUT	1000	/* 1 sec */
#define REL_MAX_RC	5

#define DHCP6_DURATION_INFINITE 0xffffffff
#define DHCP6_DURATION_MIN 30

#define DHCP6_RELAY_MULTICAST_HOPS 32
#define DHCP6_RELAY_HOP_COUNT_LIMIT 32

#define DHCP6_IRT_DEFAULT 86400	/* 1 day */
#define DHCP6_IRT_MINIMUM 600

/* DUID: DHCP unique Identifier */
struct duid {
	size_t duid_len;	/* length */
	char *duid_id;		/* variable length ID value (must be opaque) */
};

struct dhcp6_vbuf {		/* generic variable length buffer */
	int dv_len;
	caddr_t dv_buf;
};

/* option information */
struct dhcp6_ia {		/* identity association */
	u_int32_t iaid;
	u_int32_t t1;
	u_int32_t t2;
};

struct dhcp6_prefix {		/* IA_PA */
	u_int32_t pltime;
	u_int32_t vltime;
	struct in6_addr addr;
	int plen;
};

struct dhcp6_statefuladdr {	/* IA_NA */
	u_int32_t pltime;
	u_int32_t vltime;
	struct in6_addr addr;
};

/* Internal data structure */
typedef enum { DHCP6_LISTVAL_NUM = 1,
	       DHCP6_LISTVAL_STCODE, DHCP6_LISTVAL_ADDR6,
	       DHCP6_LISTVAL_IAPD, DHCP6_LISTVAL_PREFIX6,
	       DHCP6_LISTVAL_IANA, DHCP6_LISTVAL_STATEFULADDR6,
	       DHCP6_LISTVAL_VBUF
} dhcp6_listval_type_t;
TAILQ_HEAD(dhcp6_list, dhcp6_listval);
struct dhcp6_listval {
	TAILQ_ENTRY(dhcp6_listval) link;

	dhcp6_listval_type_t type;

	union {
		int uv_num;
		u_int16_t uv_num16;
		struct in6_addr uv_addr6;
		struct dhcp6_prefix uv_prefix6;
		struct dhcp6_statefuladdr uv_statefuladdr6;
		struct dhcp6_ia uv_ia;
		struct dhcp6_vbuf uv_vbuf;
	} uv;

	struct dhcp6_list sublist;
};
#define val_num uv.uv_num
#define val_num16 uv.uv_num16
#define val_addr6 uv.uv_addr6
#define val_ia uv.uv_ia
#define val_prefix6 uv.uv_prefix6
#define val_statefuladdr6 uv.uv_statefuladdr6
#define val_vbuf uv.uv_vbuf

struct dhcp6_optinfo {
	struct duid clientID;	/* DUID */
	struct duid serverID;	/* DUID */

	int rapidcommit;	/* bool */
	int pref;		/* server preference */
	int32_t elapsed_time;	/* elapsed time (from client to server only) */
	int64_t refreshtime;	/* info refresh time for stateless options */

	struct dhcp6_list iapd_list; /* list of IA_PD */
	struct dhcp6_list iana_list; /* list of IA_NA */
	struct dhcp6_list reqopt_list; /* options in option request */
	struct dhcp6_list stcode_list; /* status code */
	struct dhcp6_list sip_list; /* SIP server list */
	struct dhcp6_list sipname_list; /* SIP domain list */
	struct dhcp6_list dns_list; /* DNS server list */
	struct dhcp6_list dnsname_list; /* Domain Search list */
	struct dhcp6_list ntp_list; /* NTP server list */
	struct dhcp6_list prefix_list; /* prefix list */
	struct dhcp6_list nis_list; /* NIS server list */
	struct dhcp6_list nisname_list; /* NIS domain list */
	struct dhcp6_list nisp_list; /* NIS+ server list */
	struct dhcp6_list nispname_list; /* NIS+ domain list */
	struct dhcp6_list bcmcs_list; /* BCMC server list */
	struct dhcp6_list bcmcsname_list; /* BCMC domain list */

	struct dhcp6_vbuf relay_msg; /* relay message */
#define relaymsg_len relay_msg.dv_len
#define relaymsg_msg relay_msg.dv_buf

	struct dhcp6_vbuf ifidopt; /* Interface-id */
#define ifidopt_len ifidopt.dv_len
#define ifidopt_id ifidopt.dv_buf

	u_int authflags;
#define DHCP6OPT_AUTHFLAG_NOINFO	0x1
	int authproto;
	int authalgorithm;
	int authrdm;
	/* the followings are effective only when NOINFO is unset */
	u_int64_t authrd;
	union {
		struct {
			u_int32_t keyid;
			struct dhcp6_vbuf realm;
			int offset; /* offset to the HMAC field */
		} aiu_delayed;
		struct {
			int type;
			int offset; /* offset to the HMAC field */
			char val[16]; /* key value */
		} aiu_reconfig;
	} authinfo;
#define delayedauth_keyid authinfo.aiu_delayed.keyid
#define delayedauth_realmlen authinfo.aiu_delayed.realm.dv_len
#define delayedauth_realmval authinfo.aiu_delayed.realm.dv_buf
#define delayedauth_offset authinfo.aiu_delayed.offset
#define reconfigauth_type authinfo.aiu_reconfig.type
#define reconfigauth_offset authinfo.aiu_reconfig.offset
#define reconfigauth_val authinfo.aiu_reconfig.val
};

/* DHCP6 base packet format */
struct dhcp6 {
	union {
		u_int8_t m;
		u_int32_t x;
	} dh6_msgtypexid;
	/* options follow */
} __attribute__ ((__packed__));
#define dh6_msgtype	dh6_msgtypexid.m
#define dh6_xid		dh6_msgtypexid.x
#define DH6_XIDMASK	0x00ffffff

/* DHCPv6 relay messages */
struct dhcp6_relay {
	u_int8_t dh6relay_msgtype;
	u_int8_t dh6relay_hcnt;
	struct in6_addr dh6relay_linkaddr; /* XXX: badly aligned */
	struct in6_addr dh6relay_peeraddr; /* ditto */
	/* options follow */
} __attribute__ ((__packed__));

/* options */
#define DH6OPT_CLIENTID	1
#define DH6OPT_SERVERID	2
#define DH6OPT_IA_NA 3
#define DH6OPT_IA_TA 4
#define DH6OPT_IAADDR 5
#define DH6OPT_ORO 6
#define DH6OPT_PREFERENCE 7
#  define DH6OPT_PREF_UNDEF -1
#  define DH6OPT_PREF_MAX 255
#define DH6OPT_ELAPSED_TIME 8
#  define DH6OPT_ELAPSED_TIME_UNDEF -1
#define DH6OPT_RELAY_MSG 9
/* #define DH6OPT_SERVER_MSG 10: deprecated */
#define DH6OPT_AUTH 11
#  define DH6OPT_AUTH_PROTO_DELAYED 2
#  define DH6OPT_AUTH_RRECONFIGURE 3
#  define DH6OPT_AUTH_ALG_HMACMD5 1
#define DH6OPT_UNICAST 12
#define DH6OPT_STATUS_CODE 13
#  define DH6OPT_STCODE_SUCCESS 0
#  define DH6OPT_STCODE_UNSPECFAIL 1
#  define DH6OPT_STCODE_NOADDRSAVAIL 2
#  define DH6OPT_STCODE_NOBINDING 3
#  define DH6OPT_STCODE_NOTONLINK 4
#  define DH6OPT_STCODE_USEMULTICAST 5
#  define DH6OPT_STCODE_NOPREFIXAVAIL 6

#define DH6OPT_RAPID_COMMIT 14
#define DH6OPT_USER_CLASS 15
#define DH6OPT_VENDOR_CLASS 16
#define DH6OPT_VENDOR_OPTS 17
#define DH6OPT_INTERFACE_ID 18
#define DH6OPT_RECONF_MSG 19

#define DH6OPT_SIP_SERVER_D 21
#define DH6OPT_SIP_SERVER_A 22
#define DH6OPT_DNS 23
#define DH6OPT_DNSNAME 24
#define DH6OPT_IA_PD 25
#define DH6OPT_IA_PD_PREFIX 26
#define DH6OPT_NIS_SERVERS 27
#define DH6OPT_NISP_SERVERS 28
#define DH6OPT_NIS_DOMAIN_NAME 29
#define DH6OPT_NISP_DOMAIN_NAME 30
#define DH6OPT_NTP 31
#define DH6OPT_REFRESHTIME 32
 #define DH6OPT_REFRESHTIME_UNDEF -1
#define DH6OPT_BCMCS_SERVER_D 33
#define DH6OPT_BCMCS_SERVER_A 34
#define DH6OPT_GEOCONF_CIVIC 36
#define DH6OPT_REMOTE_ID 37
#define DH6OPT_SUBSCRIBER_ID 38
#define DH6OPT_CLIENT_FQDN 39

/* The followings are KAME specific. */

struct dhcp6opt {
	u_int16_t dh6opt_type;
	u_int16_t dh6opt_len;
	/* type-dependent data follows */
} __attribute__ ((__packed__));

/* DUID type 1 */
struct dhcp6opt_duid_type1 {
	u_int16_t dh6_duid1_type;
	u_int16_t dh6_duid1_hwtype;
	u_int32_t dh6_duid1_time;
	/* link-layer address follows */
} __attribute__ ((__packed__));

/* Status Code */
struct dhcp6opt_stcode {
	u_int16_t dh6_stcode_type;
	u_int16_t dh6_stcode_len;
	u_int16_t dh6_stcode_code;
} __attribute__ ((__packed__));

/*
 * General format of Identity Association.
 * This format applies to Prefix Delegation (IA_PD) and Non-temporary Addresses
 * (IA_NA)
 */
struct dhcp6opt_ia {
	u_int16_t dh6_ia_type;
	u_int16_t dh6_ia_len;
	u_int32_t dh6_ia_iaid;
	u_int32_t dh6_ia_t1;
	u_int32_t dh6_ia_t2;
	/* sub options follow */
} __attribute__ ((__packed__));

/* IA Addr */
struct dhcp6opt_ia_addr {
	u_int16_t dh6_ia_addr_type;
	u_int16_t dh6_ia_addr_len;
	struct in6_addr dh6_ia_addr_addr;
	u_int32_t dh6_ia_addr_preferred_time;
	u_int32_t dh6_ia_addr_valid_time;
} __attribute__ ((__packed__));

/* IA_PD Prefix */
struct dhcp6opt_ia_pd_prefix {
	u_int16_t dh6_iapd_prefix_type;
	u_int16_t dh6_iapd_prefix_len;
	u_int32_t dh6_iapd_prefix_preferred_time;
	u_int32_t dh6_iapd_prefix_valid_time;
	u_int8_t dh6_iapd_prefix_prefix_len;
	struct in6_addr dh6_iapd_prefix_prefix_addr;
} __attribute__ ((__packed__));

/* Authentication */
struct dhcp6opt_auth {
	u_int16_t dh6_auth_type;
	u_int16_t dh6_auth_len;
	u_int8_t dh6_auth_proto;
	u_int8_t dh6_auth_alg;
	u_int8_t dh6_auth_rdm;
	u_int8_t dh6_auth_rdinfo[8];
	/* authentication information follows */
} __attribute__ ((__packed__));

enum { DHCP6_AUTHPROTO_UNDEF = -1, DHCP6_AUTHPROTO_DELAYED = 2,
       DHCP6_AUTHPROTO_RECONFIG = 3 };
enum { DHCP6_AUTHALG_UNDEF = -1, DHCP6_AUTHALG_HMACMD5 = 1 };
enum { DHCP6_AUTHRDM_UNDEF = -1, DHCP6_AUTHRDM_MONOCOUNTER = 0 };

#endif /*__DHCP6_H_DEFINED*/
